### PR TITLE
Add total token count and intermittent tokens consumed so far

### PR DESCRIPTION
**Pull request by pudepiedj**
This adds the 'total of all tokens' alongside the 'total time' at the end of the log display. Example:
```
llama_print_timings:        load time =    4334.62 ms
llama_print_timings:      sample time =      19.75 ms /   235 runs   (    0.08 ms per token, 11898.73 tokens per second)
llama_print_timings: prompt eval time =     159.21 ms /     8 tokens (   19.90 ms per token,    50.25 tokens per second)
llama_print_timings:        eval time =    6892.57 ms /   234 runs   (   29.46 ms per token,    33.95 tokens per second)
llama_print_timings:       total time =    7104.83 ms /   477 tokens
```

It also adds a new command-line parameter `-stc N` or `--show_token_count N` which displays how many tokens of the context have been consumed at a user-specified interval of `N` tokens, so `-stc 512` will show `Tokens consumed so far n_past / n_ctx` (in red) every 512 tokens. (512 is the default; maybe it should be much larger to silence the report altogether; otherwise silence by making `N > ctx` in the code?) Example output from `mixtral-8x7b-instruct-v0.1.Q_K_M.gguf` with `-stc 256`:
```
This will push your local changes to a branch called `branch_name` in your forked repository
**Tokens consumed so far = 256 / 4096** 
 that is hosted on the remote named `origin`.
```
There are almost certainly neater ways to display this information, but this is a first effort.

_Motivation_: long conversations need some sense of 'how far into this have we come?' that doesn't entail reading the logs - even if the ingenious `-n -1` allows endless interactions - because of the way the earlier tokens are binned once the `-c N` limit is reached.

### CODE REVIEW

File: examples/main/main.cpp [hunk 1]
```diff
@@ -650,12 +650,16 @@ int main(int argc, char ** argv) {
...
+                if (n_past % params.token_interval == 0) {
+                    printf("\n\033[31mTokens consumed so far = %d / %d \033[0m\n", n_past, n_ctx);
+                }
             }
 
             if (!embd.empty() && !path_session.empty()) {
                 session_tokens.insert(session_tokens.end(), embd.begin(), embd.end());
                 n_session_consumed = session_tokens.size();
-            }
+                }
```
Comment (ggerganov) [inline examples/main/main.cpp:662]:
```suggestion
            }
```

File: llama.cpp [hunk 2]
```diff
@@ -10921,7 +10921,7 @@ void llama_print_timings(struct llama_context * ctx) {
             __func__, timings.t_p_eval_ms, timings.n_p_eval, timings.t_p_eval_ms / timings.n_p_eval, 1e3 / timings.t_p_eval_ms * timings.n_p_eval);
     LLAMA_LOG_INFO("%s:        eval time = %10.2f ms / %5d runs   (%8.2f ms per token, %8.2f tokens per second)\n",
             __func__, timings.t_eval_ms, timings.n_eval, timings.t_eval_ms / timings.n_eval, 1e3 / timings.t_eval_ms * timings.n_eval);
-    LLAMA_LOG_INFO("%s:       total time = %10.2f ms\n", __func__, (timings.t_end_ms - timings.t_start_ms));
+    LLAMA_LOG_INFO("%s:       total time = %10.2f ms / %5d tokens\n", __func__, (timings.t_end_ms - timings.t_start_ms), (timings.n_sample + timings.n_p_eval + timings.n_eval));
```
Comment (ggerganov) [inline llama.cpp:10924]: This should be just `timings.n_p_eval + timings.n_eval`

**Review by ggerganov — CHANGES_REQUESTED** [review]

File: common/common.h [hunk 3]
```diff
@@ -66,6 +66,7 @@ struct gpt_params {
     int32_t grp_attn_w                      = 512;   // group-attention width
     float   rope_freq_base                  = 0.0f;  // RoPE base frequency
     float   rope_freq_scale                 = 0.0f;  // RoPE frequency scaling factor
+    int32_t token_interval                  = 512;   // show token count every 512 tokens
```
Comment (ggerganov) [inline common/common.h:69]: Minor : move this after `grp_attn_w`. The idea is to not split the rope related parameters (rope, yarn) from each other

**Review by ggerganov — APPROVED** [review]

**Comment by pudepiedj** [comment]
@ggerganov Did I break something?
The latest merge won't seem to compile because of an error in `main.cpp` line 885:
```
llama.cpp/examples/main/main.cpp:885:13: error: 'break' statement not in loop or switch statement
            break;
```
This isn't where I was working so not able to see what is wrong but it look as if a `}` has gone missing.

**Comment by ggerganov** [comment]
It builds on my end - think everything is OK on `master`

**Comment by jxy** [comment]
Can we at least have the `Tokens consumed so far` print to stderr, or via LOG, so it's not mixed with and breaking the output? With `--log-disable` there's no other irrelevant text from `main` except this message coming out of the blue, or red in this case. Especially annoying in the middle of a word.

**Comment by jxy** [comment]
and the parameter name `token_interval` is unintuitive, since it really only controls that one line log that interrupts the printing of generated tokens every `token_interval` tokens.

**Comment by ggerganov** [comment]
Are there any drawbacks of what is currently on master (p.s. I fixed the `-stc`).
Initially, I didn't notice that the print is enabled by default - we don't want this, so it is now enabled only when `-ptc` > 0

**Comment by pudepiedj** [comment]
> Are there any drawbacks of what is currently on master (p.s. I fixed the `-stc`). Initially, I didn't notice that the print is enabled by default - we don't want this, so it is now enabled only when `-ptc` > 0

I agree it's better to be off by default but I didn't see how to do it, so thanks.

I think the updating of the tokens consumed is most useful when in interactive or chatml mode and that makes putting it immediately before the user prompt `> ` the best slot because it doesn't break the flow of the output. That doesn't bother me, but it clearly bothers some people, so my suggestion would solve both issues at once.

Incidentally, have you changed the 'sliding window' approach to context length when `-n -1` is used because it used to reset in the sense that `4096 / 4096` would go to something like `2048 / 4096` but now it just seems to terminate with a `zsh abort` once the context length is reached. Maybe just a glitch at my end.

Very impressed with `mixtral-8x7b-instruct-v0.1.Q_K_M.gguf` in interactive mode. Fast and reliable and great using file input to summarise as part of the prompt. Thanks for this - and everything else!

**Comment by ggerganov** [comment]
> Incidentally, have you changed the 'sliding window' approach to context length when -n -1 is used because it used to reset in the sense that 4096 / 4096 would go to something like 2048 / 4096 but now it just seems to terminate with a zsh abort once the context length is reached. Maybe just a glitch at my end.

There were changes near this functionality that could have caused a regression, but it seems to work correctly on my end. Which command are you using?

**Comment by pudepiedj** [comment]
> There were changes near this functionality that could have caused a regression, but it seems to work correctly on my end. Which command are you using?

It is certainly not ending gracefully with `-n -1` but it is always possible that I have done something to cause that. Let me try on a clean clone of the repo to see. This is the end of a `-c 3072` conversation:
```
Tokens consumed so far = 3072 / 3072 
 theirGGML_ASSERT: /Users/edsilm2/llama.cpp/ggml-metal.m:2215: false
zsh: abort      ./bin/main -m ../models/Mixtral-8x7b/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf -
```
The command was
```
make -j && ./bin/main -m ../models/Mixtral-8x7b/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf -c 3072 -ngl 99 -n -1 -s 1 -i -ins -ctk q8_0 --override-kv llama.expert_used_count=int:3 -ptc 512 --color -f '/absolute/path/to/whisper.cpp/USfilestranscribed/US1126.txt'
```

**Comment by pudepiedj** [comment]
> Context shifting with quantized KV cache (`-ctk q8_0`) is currently not supported

Ah! That explains it, but it doesn't explain why yesterday it seemed to be doing exactly that despite still being run with `-ctk q8_0`. Anyway, thank you for investigating. It certainly isn't doing it now, even with a clean clone:
```
Tokens consumed so far = 4096 / 4096 
 argGGML_ASSERT: /Users/edsilm2/llama.cpp/ggml-metal.m:2215: false
zsh: abort      ./bin/main -m ../models/Mixtral-8x7b/mixtral-8x7b-instruct-v0.1.Q4_K_M.gguf -
```

**Pull request merged**